Multi-metric display on model bias screen (#1273)

alexcreasy opened the pull request:
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->
Closes #1273 
## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits have meaningful messages (squashes happen on merge by the bot).
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work
- [ ] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

DaoDaoNoCode commented on frontend/src/pages/modelServing/screens/metrics/BiasTab.tsx:
```diff
@@ -1,38 +1,90 @@
+    `${SELECTED_CHARTS_STORAGE_KEY_PREFIX}-${inferenceService}`,
+    [],
+    true,
+    true,
```
Any reason we are going to make this a session storage but not local storage?

alexcreasy replied:
Vince asked for it to persist for the length of the session. I personally agree with this take, I think if we want to persist the graph state permanently for the user, that should be handled with a backend data store. That means any stale data would get cleaned up for us automatically and also the charts state persists across machines etc. I think session is the right balance for this, the user doesn't lose their state when they refresh, or navigate to a different inference model, which is a real nuisance.

DaoDaoNoCode commented on frontend/src/pages/modelServing/screens/metrics/BiasTab.tsx:
```diff
@@ -1,38 +1,90 @@
+          leftToolbarItem={
+            <ToolbarItem>
+              <ToolbarItem variant="label">Metrics to display</ToolbarItem>
+              <ToolbarItem>
```
Do we really need this `ToolbarItem` Wrapper here?

DaoDaoNoCode commented on frontend/src/pages/modelServing/screens/metrics/BiasTab.tsx:
```diff
@@ -1,38 +1,90 @@
+        <MetricsPageToolbar
+          leftToolbarItem={
+            <ToolbarItem>
+              <ToolbarItem variant="label">Metrics to display</ToolbarItem>
```
I am not a fan of nested `ToolbarItem`, maybe you should make them a `ToolbarGroup` and use `Stack` inside it for the `ToolbarItem`. Some style nits, anyway.

DaoDaoNoCode commented on frontend/src/pages/modelServing/screens/metrics/InferenceGraphs.tsx:
```diff
@@ -27,12 +27,12 @@ const InferenceGraphs: React.FC = () => {
+              //   const newPoint = per100(point);
+              //   const y = Math.floor(newPoint.y / (Math.floor(Math.random() * 2) + 2));
+              //   return { ...newPoint, y };
+              // },
```
Remove these comments?

DaoDaoNoCode commented on frontend/src/pages/modelServing/screens/metrics/MetricsPageToolbar.tsx:
```diff
@@ -1,28 +1,30 @@
       <ToolbarContent>
-        <ToolbarItem>
+        {leftToolbarItem && leftToolbarItem}
+        <ToolbarItem alignment={{ default: 'alignRight' }}>
```
Also, maybe consider change this to `ToolbarGroup` and stack the items inside it and give it a gutter.

DaoDaoNoCode commented on frontend/src/pages/modelServing/screens/metrics/BiasMetricConfigSelector.tsx:
```diff
@@ -0,0 +1,114 @@
+      <span id={elementId} hidden>
+        Select the metrics to display charts for
+      </span>
+      <Select
```
Is it possible to limit the width of this Select component? The layout of the toolbar will be broken if we choose a lot of selections and show them all:

<img width="1492" alt="Screenshot 2023-06-09 at 4 43 03 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/37624318/f3d5d6c1-947d-4b6c-bfcd-615ebef8deb4">

alexcreasy replied:
Really good point, I've limited it to 600px now, that seems like a good compromise, let me know what you think.

DaoDaoNoCode reviewed (changes requested):
The changes look great, just left some comments about the code quality and some nits and thoughts.
One more question, is it possible to make the page title, tabs, and toolbar sticky and only the charts section scrollable? Currently, the whole page is scrollable. If there are a lot of charts, it will be inconvenient for the users to change the toolbar settings. Just a thought, we can leave it for now but if it's not hard to implement, please change that. Or you can create a clean-up ticket and we can leave it there to fix it together.

alexcreasy commented on frontend/src/api/prometheus/serving.ts:
```diff
@@ -19,6 +23,7 @@ export const useModelServingMetrics = (
   timeframe: TimeframeTitle,
   lastUpdateTime: number,
   setLastUpdateTime: (time: number) => void,
+  refreshInterval: RefreshIntervalTitle,
```
good catch, I meant to come back and do this after I got it working it working for bias metrics, whoops :)

alexcreasy replied:
Done

alexcreasy commented on frontend/src/pages/modelServing/screens/metrics/BiasMetricChartWrapper.scss:
```diff
@@ -0,0 +1,9 @@
+}
+
+.dashboard-expandable-section-heading > .pf-c-expandable-section__toggle:hover > * {
+  color: var(--pf-global--icon--Color--dark) !important;
```
Thanks, Vince requested it match a specific OpenShift console item, so had to do a bit of digging to get the matching PF vars and getting a specific enough selector to override the default heh :)

alexcreasy commented on frontend/src/pages/modelServing/screens/metrics/BiasMetricChartWrapper.scss:
```diff
@@ -0,0 +1,9 @@
+.dashboard-expandable-section-heading > .pf-c-expandable-section__toggle > * {
```
Good shout, done :)

alexcreasy commented on frontend/src/pages/modelServing/screens/metrics/MetricsPageToolbar.tsx:
```diff
@@ -1,28 +1,30 @@
     <Toolbar>
       <ToolbarContent>
-        <ToolbarItem>
+        {leftToolbarItem && leftToolbarItem}
```
JSX handles nullish values really well, I keep forgetting this :)

alexcreasy commented on frontend/src/pages/modelServing/screens/metrics/BiasMetricConfigSelector.tsx:
```diff
@@ -0,0 +1,114 @@
+
+  return (
+    <>
+      <span id={elementId} hidden>
```
This was part of the PF example I used - I assume it's an accessibility feature as it's referenced in the "aria-labelledby" property ?

Happy to remove if it shouldn't be there.

alexcreasy commented on frontend/src/pages/modelServing/screens/metrics/BiasMetricConfigSelector.tsx:
```diff
@@ -0,0 +1,114 @@
+          if (isBiasSelectOption(item)) {
+            if (selected.find(byId(item))) {
+              // User has de-selected an item.
+              changeState([...selected.filter(byNotId(item))]);
```
good spot!

alexcreasy commented:
@DaoDaoNoCode I've addressed all your comments, let me know WYT :)

I had a quick go at implementing the sticky toolbar - the PF examples didn't work - I imagine the parent container properties will need some playing with to get it working. As this wasn't an explicit requirement - or explicitly requested by UXD, I'm going to put this under the "nice to have" category and we can try and add it if we have any spare time at the end (or possibly in explainability phase 2).

Thanks for the detailed review.

DaoDaoNoCode reviewed (approved):
lgtm!

alexcreasy commented:
/approve
